pkg/deploy/deployment_che.go

```diff
@@ -127,9 +128,10 @@ func NewCheDeployment(cr *orgv1.CheCluster, cheImage string, cheTag string, cmRe
 										Scheme: corev1.URISchemeHTTP,
 									},
 								},
-								InitialDelaySeconds: 50,
+								InitialDelaySeconds: 200, // Start liveness probe after readiness probe has finished
```

Comment:
> Start liveness probe after readiness probe has finished

The comment is misleading IMO. `initialDelaySeconds: Number of seconds after the container has started before liveness or readiness probes are initiated.`

Reply:
You're right, the comment is misleading and not precise enough.

What I was meaning is that we don't want the liveness probe (that would *restart* the POD in case of failure) to start checking (just after POD start) before the POD is, for the first time, either declared __*ready*__ or __*failed*__ thanks to the readiness probe work.

That's important because liveness probe and readiness probe **use the same check**.

But **after** the POD is started and has been declared as __*ready*__ by the readiness probe, then the liveness probe will starts checking, and, since it is more requiring than the readiness probe, we can be sure that it will restart the POD if any unavailability occurs.

Reply:
I do not really follow the relation of liveness probe to readiness.
I mean I understand and agree on the following formulating
```
What I was meaning is that we don't want the liveness probe (that would restart the POD in case of failure) to start checking (just after POD start) before the POD is
```
before tomcat inside of POD is started. 
But I don't understand how `ready`/`failed` readiness probe influence liveness probe at all.

Also, note in case of 
```
FailureThreshold:    10,
TimeoutSeconds:      5,
```
if tomcat is hanging or crashed, K8s/OpenShift will continue set client traffic to it for 50 seconds (PeriodSeconds*FailureThreshold). Do we really need it?

What if we apply the following changes:
```
readiness:
  // Che can become available in minimum 25 seconds
  // If it's not available within 75 seconds, it will be marked as **Not Ready**
  // and check will continue to happen each 10 seconds
  // it will become ready after first successfull probe
  // P.S. I do know if it's too bad that we can have unready starting tomcat in such case instead 
  // of just starting but it allows Che Server process traffic faster if it's already ready, no need to wait 60 seconds.
  InitialDelaySeconds: 25, // maybe like 40 seconds
  FailureThreshold:    5,
  PeriodSeconds:       10,
liveness:
  // it's a normal case if tomcat start takes ~ 200 seconds
  // do not check liveness before to prevent killing Che Server that is still starting
  InitialDelaySeconds: 200,
  PeriodSeconds:       10,
  FailureThreshold:    3,
  ...
```

Reply:
> Also, note in case of
> 
> ```
> FailureThreshold:    10,
> TimeoutSeconds:      5,
> ```
> 
> if tomcat is hanging or crashed(in the process of work, after successful start), K8s/OpenShift will continue set client traffic to it for 50 seconds (PeriodSeconds*FailureThreshold). Do we really need it?

No not when the liveness probe will be running afaict. So are you speaking of the case when Tomcat would be successfully started after 60 seconds, the container would be marked as `Ready` by the readiness probe, and the tomcat server would fail again just after ... Then we would still send traffic to it during 140 seconds (200 - 60), until liveness probe would be initiated and would detect the crashed tomcat and restart the POD ?

Reply:
> No not when the liveness probe will be running afaict. So are you speaking of the case when Tomcat would be successfully started after 60 seconds, the container would be marked as Ready by the readiness probe, and the tomcat server would fail again just after ... Then we would still send traffic to it during 140 seconds (200 - 60), until liveness probe would be initiated and would detect the crashed tomcat and restart the POD ?

Let me clarify the scenario I would like to highlight:
Che Server is started, it works for about 10 minutes, both of readiness and liveness probes are done successfully but then let's imagine connection pool is full and tomcat refused every request.
In our case, readiness and liveness probes are the same and with current values
1. liveness probe: period 10, FailureThreshold 3;  <- container will be restarted after 30 seconds;
2. readiness probe: period 10, FailureThreshold 10; <- container will receive all traffic until it won't be restarted, because 100 seconds it's needed for K8s to mark it as *Not Ready*.

In my understanding, readiness probe should fail sooner than lineness, it's a case then my application is temporary unavailable but let's give some time it to recover itself.

See https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-readiness-probes

Reply:
>```
>  // If it's not available within 75 seconds, it will be marked as **Not Ready**
>  // and check will continue to happen each 10 seconds
>  // it will become ready after first successful probe
>```

Yes, you're right, But then after the 75 seconds we will have `Unhealthy` container events occurring in the POD events, which we probably don't want if we consider this time to be still acceptable to start Che.